### DISCUSSION TITLE
DateTime to String Converter

### DISCUSSION THREAD

**Discussion by ziomek64** [CommunityToolkit/Maui#125] [New Feature Discussions]
Hi, 
I'd like to make a proposal of DateTime to String converter. Why should someone need such converter? Well, you can bind Text to DateTime and use formattedstring, but it might not allow you to do stuff like AM/PM distinction (from what I've tried). This is not much, but someone might want it :) 

I've personally needed it so I've made some rough code and it works so I thought I might contribute. It accepts all formats mentioned in this page https://docs.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=net-5.0 as a parameter. The code probably needs a styling such as you use. I made it for Forms though but I noticed you don't accept new contributions there or did that change since maui got pushed? I could try to make it for MAUI or for Forms if you think i can do it there.

**Comment by TheCodeTraveler** [comment]
Sounds like a cool idea! Do you have a API spec or a implementation you could share?

I'd also love to see a way to include/handle `string` and `IFormatProvider` in these APIs 
- `ToString(String, IFormatProvider)`
- `ToString(String)`
- `ToString(IFormatProvider)`

**Comment by ziomek64** [comment]
Please take a look at my sample repo - https://github.com/ziomek64/Converters  , bear in mind that is a Forms project.  Simple page with code behind and a converter.

I haven't touched the code since my initial post. Let me know what you think and how exactly would you like to see those APIs handled after having a look.

**Reply by TheCodeTraveler** [reply]
Thanks @ziomek64! Can you copy/paste the implementation to this Discussion here? 

I'd like to ensure we have all of the details documented here in case something happens to an external repo 👍

**Reply by ziomek64** [reply]
> 
> 
> Thanks @ziomek64! Can you copy/paste the implementation to this Discussion here?
> 
> I'd like to ensure we have all of the details documented here in case something happens to an external repo 👍

Just to avoid pasting long code - unlisted pastebin https://pastebin.com/hxS4xve2 

Or do You want me to literally just paste it here?

**Reply by TheCodeTraveler** [reply]
Please paste the code to this Discussion

**Reply by ziomek64** [reply]
```
using System;
using System.Globalization;
using System.Linq;
using Xamarin.Forms;

namespace Converters
{
    /// <summary>
    ///     Converts a DateTime to string value.
    /// </summary>
    public class DateTimetoStringConverter : IValueConverter
    {
        /// <summary>
        /// Default format used when parameter is not provided.
        /// Example string with this format - 10/31/2008 5:04 PM
        /// </summary>
        private const string DefaultFormat = "G";
        
        /// <summary>
        /// List of compatible formats.
        /// See https://docs.microsoft.com/en-us/dotnet/api/system.datetime.tostring?view=net-5.0 for more information.
        /// </summary>
        private static readonly string[] Formats = {"d", "D", "f", "F", "g", "G", "m", "o", "r",
            "s", "t", "T", "u", "U", "Y"};
        
        /// <param name="value">The value to convert.</param>
        /// <param name="targetType">The type of the binding target property. This is not implemented.</param>
        /// <param name="parameter">Additional parameter for the converter to handle as a format. If not provided, null or empty it is "G" format by default.</param>
        /// <param name="culture">The culture to use in the converter.</param>
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is not DateTime datetime)
                throw new ArgumentException("Value is not a valid DateTime.");

            string format;

            string parameterString = parameter?.ToString();

            if (string.IsNullOrEmpty(parameterString))
            {
                format = DefaultFormat;
            }
            else
            {
                if (!(Formats.Contains(parameterString)))
                    throw new ArgumentException("Format parameter not supported.");
                
                format = parameterString;
            }
            
            return datetime.ToString(format, culture);
        }
        
        /// <summary>
        /// This method is not implemented and will throw a <see cref="NotImplementedException"/>.
        /// </summary>
        /// <param name="value">N/A</param>
        /// <param name="targetType">N/A</param>
        /// <param name="parameter">N/A</param>
        /// <param name="culture">N/A</param>
        /// <returns>N/A</returns>
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
            => throw new NotImplementedException();
    }
}
```

**Reply by TheCodeTraveler** [reply]
Thanks! Can you share some examples how how to use it in XAML, including examples using format and culture?

**Reply by ziomek64** [reply]
That's how I use it in my test app:
```
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage
    x:Class="Converters.MainPage"
    xmlns="http://xamarin.com/schemas/2014/forms"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:converters="clr-namespace:Converters;assembly=Converters">

    <ContentPage.Resources>
        <ResourceDictionary>
            <converters:DateTimetoStringConverter x:Key="DateTimetoStringConverter" />
        </ResourceDictionary>
    </ContentPage.Resources>

    <StackLayout>
        <Label Text="{Binding date, Converter={StaticResource DateTimetoStringConverter}, ConverterParameter='D'}" />
    </StackLayout>

</ContentPage>
```
I hardcoded a parameter (format) for myself. 
Sadly I don't know if there is a way to implement culture info in XAML - https://github.com/xamarin/Xamarin.Forms/issues/12181
It uses CurrentCulture or something like that from what I've gathered - https://github.com/xamarin/Xamarin.Forms/issues/14600

**Comment by chrisstaley** [comment]
I really like this idea. A couple of comments:

1. For maximum flexibility I would suggest it also allow custom formats in addition to the built in formats. And if we're going that far, we could even just make it an object-to-string converter that accepts any custom string format.
2. Aside from that detour, it would also be nice to have the ability to control the casing of the AM/PM indicators (i.e. am/pm vs. AM/PM). I've had designers insist they be lowercase.

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889
